src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/ActivitySource.cs

```diff
@@ -0,0 +1,357 @@
+            }
+
+            if (s_allListeners.AddIfNotExist(listener))
+            {
+                s_activeSources.EnumWithAction(source => {
+                    var shouldListenTo = listener.ShouldListenTo;
+                    if (shouldListenTo != null && shouldListenTo(source))
+                    {
+                        source.AddListener(listener);
+                    }
+                });
+            }
+        }
+
+        internal void AddListener(ActivityListener listener)
+        {
+            if (_listeners == null)
+            {
+                Interlocked.CompareExchange(ref _listeners, new SynchronizedList<ActivityListener>(), null);
+            }
+
+            _listeners.AddIfNotExist(listener);
+        }
+
+        internal static void DetachListener(ActivityListener listener)
+        {
+            s_allListeners.Remove(listener);
+
+            s_activeSources.EnumWithAction(source => {
+                var listeners = source._listeners;
+                listeners?.Remove(listener);
+            });
+        }
+
+        internal void NotifyActivityStart(Activity activity)
+        {
+            Debug.Assert(activity != null);
+
+            // _listeners can get assigned to null in Dispose.
+            SynchronizedList<ActivityListener>? listeners = _listeners;
+            if (listeners != null &&  listeners.Count > 0)
+            {
+                listeners.EnumWithAction(listener => listener.ActivityStarted?.Invoke(activity));
+            }
+        }
+
+        internal void NotifyActivityStop(Activity activity)
+        {
+            Debug.Assert(activity != null);
+
+            // _listeners can get assigned to null in Dispose.
+            SynchronizedList<ActivityListener>? listeners = _listeners;
+            if (listeners != null &&  listeners.Count > 0)
+            {
+                listeners.EnumWithAction(listener => {
+                    listeners.EnumWithAction(listener => listener.ActivityStopped?.Invoke(activity));
+                });
+            }
+        }
+    }
+
+    // SynchronizedList<T> is a helper collection which ensure thread safety on the collection
+    // and allow enumerating the collection items and execute some action on the enumerated item and can detect any change in the collection
+    // during the enumeration which force restarting the enumeration again.
+    // Causion: We can have the action executed on the same item more than once which is ok in our scenarios.
+    internal class SynchronizedList<T>
+    {
+        private List<T> _list;
+        private uint _version;
+
+        public SynchronizedList() => _list = new List<T>();
+
+        public void Add(T item)
+        {
+            lock (_list)
+            {
+                _list.Add(item);
+                _version++;
+            }
+        }
+
+        public bool AddIfNotExist(T item)
+        {
+            lock (_list)
+            {
+                if (!_list.Contains(item))
```

Comment:
How big do we expect these lists to be?  And more importantly, what's the worst case?  N calls to AddIfNotExist results in O(N^2) work here.

Reply:
We are expecting the lists to be short. For activity sources, we expect would be one per component if the component publishing any tracing data. For the listeners, usually, the list will be shorter as in main scenario cases we expect one listener for who is interested to get the Activity notification (e.g. OpenTelemetry for instance). usual apps/Libraries not going to create listeners.